refactor: migrate to starflow

bepri opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/canonical/snapcraft/blob/main/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Should be reviewable on a commit-by-commit basis. Comments will be left on individual files explaining why some changes were necessary. As some words of encouragement: about 250 of the file changes were purely from `make format`!

bepri commented on .shellcheckrc:
Was meant to be temporary, was almost accidentally forever

bepri commented on TESTING.md:
Superceded by changes to `HACKING.md`

bepri commented on manual-tests.md:
Was more or less all outdated information and references to files that were no longer there

bepri commented on runtests.sh:
Outdated - referenced files that were no longer there

bepri commented on snapcraft_legacy/internal/common.py:
```diff
@@ -35,13 +35,16 @@
 _plugindir = _DEFAULT_PLUGINDIR
-_DEFAULT_SCHEMADIR = os.path.join(sys.prefix, "share", "snapcraft", "schema")
+
+_BASE_DIR = Path(__file__).parent.parent.parent
```
These changes were necessary due to the way `setup.py` used to package the `schema/`, `extensions/`, and `keyrings/` directories. It made use of the `data_files` kwarg in `setup()`, which is strongly discouraged. In the process of migrating to `pyproject.toml`, we now use "package data" and "package files" which will remain at the root of the project and are organized to the `share/` directory in `snapcraft.yaml` for the snapped version of legacy snapcraft.

lengau replied:
```suggestion
_BASE_DIR = Path(__file__).parents[2]
```

bepri commented on tools/environment-setup-local.sh:
This file (along with many of the other tools) were either outdated or superceded by the starbase makefile

bepri commented on units.py:
This file was outdated, from a time before `pytest`

lengau commented on .github/workflows/qa.yaml:
```diff
@@ -0,0 +1,24 @@
+    uses: canonical/starflow/.github/workflows/test-python.yaml@main
+    with:
+      # Snapcraft currently only tests on Python 3.12 on Ubuntu 24.04
+      fast-test-platforms: '["ubuntu-24.04"]'
```
@sergiusens should we be using self-hosted runners for this or GH-hosted ones?

lengau commented on appveyor.yml:
Do we even need this file anymore?

lengau commented on requirements-noble.txt:
```diff
@@ -0,0 +1,2 @@
+python-apt @ https://launchpad.net/ubuntu/+archive/primary/+sourcefiles/python-apt/2.7.6/python-apt_2.7.6.tar.xz \
+    --hash=sha256:19918f9f2e30f690e89bb7c2eeb8b9bc229a2a0792df3aecf632ad4cbc6ee4b9
```
```suggestion
python-apt @ https://launchpad.net/ubuntu/+archive/primary/+sourcefiles/python-apt/2.7.7ubuntu4/python-apt_2.7.7ubuntu4.tar.xz \
    --hash=sha256:9833fcccf33e0fbf00d62df93c0b6165dc00edd9f635251be4cde1a9dc708e88
```

lengau commented on pyproject.toml:
```diff
@@ -1,14 +1,197 @@
+    "craft-platforms~=0.4",
+    "craft-store>=3.0.2,<4.0.0",
+    "cryptography==43.0.3", # Pinned due to https://github.com/canonical/snapcraft/issues/5217
+    "docutils<0.20", # Frozen until we can update sphinx dependencies.
```
This looks like it actually belongs under docs dependencies.

lengau commented on pyproject.toml:
```diff
@@ -1,14 +1,197 @@
+    "cryptography==43.0.3", # Pinned due to https://github.com/canonical/snapcraft/issues/5217
+    "docutils<0.20", # Frozen until we can update sphinx dependencies.
+    "gnupg",
+    "jsonschema==2.5.1",
```
This was pinned way back in 2021: https://github.com/canonical/snapcraft/commit/cbae1226406e09ace71d83fb534e174531ddba7f

We should probably update it.

mr-cal commented:
The spread test failures look like inconsequential changes to yaml formatting, so those should be quick fixes.

mr-cal commented on schema/snapcraft.json:
Is this new content or the same content but formatted?

mr-cal commented on snap/snapcraft.yaml:
```diff
@@ -165,44 +163,53 @@ parts:
 
-        version="$(python3 setup.py --version)"
-        ${SNAP}/libexec/snapcraft/craftctl set version="$version"
+      version=$(PYTHONPATH=/root/parts/snapcraft/install/lib/python3.12/site-packages python3 -c "import snapcraft;print(snapcraft.__version__)")
```
```suggestion
      version=$(PYTHONPATH=$CRAFT_PART_INSTALL/lib/python3.12/site-packages python3 -c "import snapcraft;print(snapcraft.__version__)")
```

mr-cal commented on snapcraft_legacy/internal/mountinfo.py:
```diff
@@ -58,9 +58,7 @@ def __init__(self, *, mountinfo_file: str = "/proc/self/mountinfo") -> None:
-        root_mounts = collections.defaultdict(
-            list
-        )  # type: Dict[str, List[Mount]]  # noqa
+        root_mounts = collections.defaultdict(list)  # type: Dict[str, List[Mount]]  # noqa
```
lol type comments

mr-cal commented on tests/spread/core24-suites/scriptlets/scriptlet-failures/task.yaml:
```diff
@@ -1,6 +1,7 @@
 summary: Validate scriptlet failures
 
-restore: |  # If this next line fails, delete it and replace it with the one following it
+restore:
+  | # If this next line fails, delete it and replace it with the one following it
```
Can you drop this outdated comment and move the pipe symbol back to where it belongs?

This seems like a bug in the formatter.

mr-cal replied:
Now out of scope for this effort, a future enhancement will be to not ignore yaml files in test/spread

mr-cal replied:
Now out of scope for this effort, a future enhancement will be to not ignore yaml files in test/spread

mr-cal commented on tests/spread/core24/set-version-twice/task.yaml:
```diff
@@ -4,7 +4,8 @@ prepare: |
 
-restore: |  # If this next line fails, delete it and replace it with the one following it
+restore:
+  | # If this next line fails, delete it and replace it with the one following it
```
Same here

mr-cal replied:
Now out of scope for this effort, a future enhancement will be to not ignore yaml files in test/spread

mr-cal reviewed (commented):
This is spooky (in a good way).  I ran `make setup` and `make test-fast` and within a few seconds I was running unit tests.

I have a few nitpicks but overall everything looks great! This will significantly improve developing and lower the barrier to entry for community contributions.

I'm wondering if this PR is a good time to merge starbase's git history into Snapcraft.  The advantage of doing this is that subsequent starbase commits can be brought into Snapcraft with small and easy `git merge`s.

lengau commented on extensions/desktop/kde-neon-6/launcher-specific:
```diff
@@ -35,7 +35,7 @@ append_dir QML2_IMPORT_PATH "$SNAP_DESKTOP_RUNTIME/lib/$ARCH"
 export QTWEBENGINEPROCESS_PATH="$SNAP_DESKTOP_RUNTIME/usr/lib/qt6/libexec/QtWebEngineProcess"
 # And QtWebEngine's path to resources
-QTWEBENGINE_RESOURCES_PATH="$SNAP_DESKTOP_RUNTIME/usr/share/qt6/resources"
+export QTWEBENGINE_RESOURCES_PATH="$SNAP_DESKTOP_RUNTIME/usr/share/qt6/resources"
```
Yeah, Shellcheck found this and I'm pretty sure this change fixes a bunch of Qt related snap issues

lengau commented on common.mk:
```diff
@@ -0,0 +1,287 @@
+	@echo ::endgroup::
+endif
+
+.PHONY: lint-docs
```
Yes. Unfortunately IDK how to override this lint-docs without getting that.

mr-cal replied:
I'm fine to leave this as-is in favor of landing the PR sooner.

lengau commented on common.mk:
```diff
@@ -0,0 +1,287 @@
+	APT := apt-get
+endif
+
+PRETTIER=npm exec --package=prettier -- prettier
```
Makes sense to me. We can also override this in `Makefile` if we want.

bepri replied:
`uv build .` from `pack-pip` also seems to flood the terminal pretty badly. How do we feel about changing it to `uv build --quiet .`? Here's an example successful lint run with that:
```
❯ make lint
Makefile:28: warning: overriding recipe for target 'lint-docs'
common.mk:166: warning: ignoring old recipe for target 'lint-docs'
ruff check  snapcraft tests
ruff format --diff  snapcraft tests
All checks passed!
190 files already formatted
uv run codespell --toml pyproject.toml  snapcraft tests
uv run mypy --show-traceback --show-error-codes snapcraft
Success: no issues found in 108 source files
npm exec --package=prettier -- prettier --log-level warn --check "**/*.{yaml,yml,json,json5,css,md}"
pyright --pythonpath .venv/bin/python
0 errors, 0 warnings, 0 informations 
git ls-files | file --mime-type -Nnf- | grep shellscript | cut -f1 -d: | xargs -r shellcheck
uv run "--group=docs" "--group=dev-oracular" sphinx-lint --ignore docs/reference/commands --ignore docs/_build --enable all docs -d missing-underscore-after-hyperlink,missing-space-in-hyperlink,line-too-long
No problems found.
uv build -q .
uv tool run twine check dist/*
Checking dist/snapcraft-8.5.1.post65+dirty-py3-none-any.whl: PASSED
Checking dist/snapcraft-8.5.1.post65+dirty.tar.gz: PASSED
```

mr-cal replied:
Looks good to me!

lengau replied:
https://github.com/canonical/starbase/pull/345

bepri commented:
> I'm wondering if this PR is a good time to merge starbase's git history into Snapcraft. The advantage of doing this is that subsequent starbase commits can be brought into Snapcraft with small and easy `git merge`s.

That's the plan! I'm leaving that rebase out of this PR until it's time to merge so that the commit history is easier on reviewers.

bepri commented:
@medubelko If you'd like, the documentation changes on this PR worth reviewing are as follows:

- Deleted `manual-tests.md`, it had a lot of outdated information.
- Migrated the little bit of useful information remaining in `TESTING.md` into `HACKING.md` and then deleted the former.
- Changed the "Code style" listed in the README to "ruff" (very minor, haha)
- Created new environment setup instructions to `HACKING.md`.

Points #2 and #4 are probably the most important to get right here. Let me know if you get any questions!

medubelko commented on Makefile:
```diff
@@ -1,55 +1,104 @@
+	$(warning Cannot install build dependencies without apt.)
+	$(warning Please ensure the equivalents to these packages are installed: $(APT_PACKAGES))
+else
+	sudo $(APT) install $(APT_PACKAGES)
```
:tada:

medubelko commented on HACKING.md:
```diff
@@ -23,78 +23,98 @@ newgrp lxd
+
+### Testing
+
+For unit testing, you can use `make` recipes:
```
I think we should push for standard testing.

```suggestion
For unit testing, use the provided make recipes:
```

medubelko commented on HACKING.md:
```diff
@@ -23,78 +23,98 @@ newgrp lxd
+### Tooling
+
+We use a large number of tools for our project. Most of these are installed for
+you with `make setup`, but you'll need to install Python 3.12 separately.
```
Should we specify APT here, too?

If `make setup` throws without it, it feels safe to say that it's a requirement, and anyone who wants to develop without it must replicate the installation on their own.

bepri replied:
I think calling apt a "requirement" feels a bit weird to me. I suppose, strictly speaking, it is a requirement to use our automated "setup" step, but it isn't strictly a requirement to (for example) run any of the tests as once your environment is already set up, you'll never touch it.

It's also worth noting that this PR doesn't introduce anything new for apt to install. There were always "ghost" system package dependencies that you just had to trial & error your way through installing in order to get all of the tests to run.

Curious what others think, though.

lengau replied:
You don't need apt, it's just that if you don't have apt you have to install the packages for your development environment yourself. So not a requirement, but a convenience. That's also why we use the warning function (not error) if we don't detect certain packages but you don't have apt.

lengau replied:
You don't need apt, it's just that if you don't have apt you have to install the packages for your development environment yourself. So not a requirement, but a convenience. That's also why we use the warning function (not error) if we don't detect certain packages but you don't have apt.

medubelko commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
 ```
 
-Import your keys (`ssh-import-id`) and add a `Host` entry to your ssh config if you are interested in [Code's](https://snapcraft.io/code) [Remote-SSH](https://code.visualstudio.com/docs/remote/ssh) plugin.
+For integration testing, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel. In order to use spread, first build Snapcraft using:
````
```suggestion
For compatibility testing across supported platforms, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel.

To test with Spread, first build Snapcraft into a snap:
```

medubelko replied:
Ignore.

medubelko commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
+go install github.com/snapcore/spread/cmd/spread@latest
+```
+
+Note: Ensure that the `$HOME/go/bin` directory is on your path after installation.
````
```suggestion
Ensure that the installation added the `$HOME/go/bin` directory to the `$PATH` environment variable.
```

medubelko commented on HACKING.md:
```diff
@@ -23,78 +23,98 @@ newgrp lxd
+
+Note: Ensure that the `$HOME/go/bin` directory is on your path after installation.
+
+Then, you can run the integration tests using a local LXD backend with:
```
```suggestion
Then, you can run the compatibility tests using a local LXD backend with:
```

medubelko replied:
Ignore.

medubelko commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
+```
 
-See the [Testing guide](TESTING.md).
+Or, you can run them in Google Cloud if you have a `SPREAD_GOOGLE_KEY`, with:
````
I think we should provide a link to a doc that describes the key – it's not evident what this is or how to get it.

medubelko replied:
Let's not do a link, then, but a quick phrase in that sentence describing it as Google Cloud credentials.

medubelko commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
+
+Then, you can run the integration tests using a local LXD backend with:
+
+```
````
```suggestion
```shell
```

medubelko commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
-### Testing
+Next, install Spread with Go:
+
+```
````
```suggestion
```shell
```

bepri commented on HACKING.md:
````diff
@@ -23,78 +23,98 @@ newgrp lxd
 ```
 
-Import your keys (`ssh-import-id`) and add a `Host` entry to your ssh config if you are interested in [Code's](https://snapcraft.io/code) [Remote-SSH](https://code.visualstudio.com/docs/remote/ssh) plugin.
+For integration testing, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel. In order to use spread, first build Snapcraft using:
````
Spread tests are super slow to run, I think encouraging them every time might make the effort required to contribute artificially high. I'm a fan of your two-section approach, what do you think @mr-cal?

mr-cal commented on snapcraft_legacy/plugins/v2/crystal.py:
```diff
@@ -96,7 +96,13 @@ def get_build_commands(self) -> List[str]:
-        env = dict(LANG="C.UTF-8", LC_ALL="C.UTF-8", SNAP=os.getenv("SNAP"), SNAP_NAME=os.getenv("SNAP_NAME"))
+        # Make sure snap-related environment variables survive through the shell call
+        snap_dict = {}
+        for key in ["SNAP", "SNAP_NAME"]:
```
```bash
 env | sort | grep "SNAP"
SNAP_ARCH=amd64
SNAP_COMMON=/var/snap/snapcraft/common
SNAP_CONTEXT=dbaQtG7ZJyq_4g-xsNR4-ISbCA2g_cyukr1Y_lMObylQD1nyqEU9
SNAP_COOKIE=dbaQtG7ZJyq_4g-xsNR4-ISbCA2g_cyukr1Y_lMObylQD1nyqEU9
SNAP_DATA=/var/snap/snapcraft/x1
SNAP_EUID=1000
SNAP_INSTANCE_KEY=
SNAP_INSTANCE_NAME=snapcraft
SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void
SNAP_NAME=snapcraft
SNAP_REAL_HOME=/home/developer
SNAP_REEXEC=
SNAP_REVISION=x1
SNAP=/snap/snapcraft/x1
SNAP_UID=1000
SNAP_USER_COMMON=/home/developer/snap/snapcraft/common
SNAP_USER_DATA=/home/developer/snap/snapcraft/x1
SNAP_VERSION=8.6.0.post1+git664560802
```

Looking at the latest failure, this may be more robust if you grab everything that begins with `SNAP`, rather than the slippery slope of adding one envvar at a time.

bepri commented:
Rebased on main and cleaned up commit history a bit

mr-cal commented on snap/snapcraft.yaml:
```diff
@@ -165,44 +163,53 @@ parts:
+
   snapcraft:
     source: .
     plugin: python
```
Snapcraft 8.6.1 is on stable with the uv fix :point_right: :point_left: 

I'm also fine if we switch to uv in a separate PR.

bepri replied:
I'll gladly do this in a separate PR! Let's keep this one as quiet as it possibly can be at this point lol

mr-cal commented on tools/staging_env.sh:
Can you keep this? I still use it

mr-cal commented:
Merging with failing tests for the following reasons:

* store and 22.04 failures are pre-existing flaky failures
* osv scanner is for cryptography, which we are handling in #5217 
* Conventional commits fail because it brings in old commits from starbase that pre-dated the check
